### ISSUE TITLE
Use function name in the SAM template generation with package command

### ISSUE THREAD

**Issue by claudiopastorini** [aws/chalice#935]
I want to deploy my chalice project using **CloudFormation**. I followed the [doc](https://chalice.readthedocs.io/en/latest/topics/cfn.html#aws-cloudformation-support) for the deploy of the app but I discovered that the `Physical ID` (the function name) of the deployed Lambdas were changed by CloudFormation.

After investigating a bit on [AWS CloudFormation Template reference](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-function.html) I discovered that if a Lambda Resource has not a name:

> AWS CloudFormation generates a unique physical ID and uses that ID for the function's name

This is annoying because I use the function name (`{app}-{stage}-{lambda_function_name}`) in other projects and with this structured name I can link it in an easy way and it is also different from the way that the `chalice deploy` command gives names to Lambdas.

In order to overcome this problem I simply add to the `lambdafunction_definition` dict:
```python
                'FunctionName': resource.function_name
```
to the `_generate_lambdafunction()` in the `package.py` file:
https://github.com/aws/chalice/blob/cb3059551da903accff7d91d710fc998adfc1289/chalice/package.py#L102-L116

I think that this is a good solution for usability (it makes easier to invoke Lambda through a simple and descriptive name) and for consistency with the `deploy` command.

If you like it I can make a PR with this simple modification, maybe, in order to maintaining backward compatibility, could be useful to leave to the user with the package command the random creation of the `Physical ID` or the use of the Chalice's function name.

**Comment by kadrach** [comment]
The general caveat of explicitly naming CloudFormation resources is that you lose the ability to perform updates on the stack that require replacement of the resource.

It looks like currently all Lambda function properties support updates without interruptions.

**Comment by claudiopastorini** [comment]
Yeah, @kadrach I read this but I did not really understand what it means. I also tried to make a change and to deploy again and it works without any problems (and it seems without interruptions).

I am new to AWS services in particular in CloudFormation and I do not really understand what does it means.

**Comment by stealthycoin** [comment]
You mean reference the function it from other projects? Its much more robust to update the CloudFormation template after it is generated by Chalice and either add the other resource you need to the one template, or if you want to keep them separate you can inject an [output into the template](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/outputs-section-structure.html) and then reference that from other stacks. 

I like using troposphere have a little post-processing script that takes the template and injects extra resources into them (or updates them).

**Comment by stealthycoin** [comment]
For example something like this:

```python
import sys
import json
import argparse

from troposphere import Output
from troposphere import Export
from troposphere.template_generator import TemplateGenerator


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        'template_location', metavar='TEMPLATE_LOCATION',
        help=(
            'The path to the template to be augmented. '
            'The template will be overwritten.'
        )
    )
    parser.set_defaults(func=_run)

    args = parser.parse_args()
    rc = args.func(args)
    sys.exit(rc)


def _run(args):
    template_location = args.template_location
    template = _load_template(template_location)
    _inject_outputs(template)
    _write_template(template_location, template)
    return 0


def _inject_outputs(template):
    function_to_export = template.resources['FooFunction']
    template.add_output(Output(
        'ExportFooFunction',
        Description='FooFunction that returns hello world.',
        Export=Export('ChaliceFooFunction'),
        Value=function_to_export.ref(),
    ))


def _load_template(path):
    with open(path, 'r') as f:
        data = json.loads(f.read())
    template = TemplateGenerator(data)
    return template


def _write_template(path, template):
    json_template = template.to_json()
    print(json_template)
    with open(path, 'w') as f:
        f.write(json_template)


if __name__ == '__main__':
    sys.exit(main())
```

And I can have a script that builds and deploys that looks like this:
```
#!/bin/sh
chalice package /tmp/packaged
python augment.py /tmp/packaged/sam.json
aws cloudformation package --template-file /tmp/packaged/sam.json --s3-bucket <BUCKET> --output-template /tmp/packaged/generated.yml
aws cloudformation deploy --template-file /tmp/packaged/generated.yml --capabilities CAPABILITY_IAM --stack-name FooStack

```

**Comment by claudiopastorini** [comment]
> You mean reference the function it from other projects? Its much more robust to update the CloudFormation template after it is generated by Chalice and either add the other resource you need to the one template, or if you want to keep them separate you can inject an [output into the template](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/outputs-section-structure.html) and then reference that from other stacks.

Yes, I need to use my functions in other projects and so I use a simple name in order to reference them. As I understood I can reference my resources from outside in a simple way through `Output`, isn'it? 

```python
function_to_export = template.resources['FooFunction']
template.add_output(Output(
        'ExportFooFunction',
        Description='FooFunction that returns hello world.',
        Export=Export('ChaliceFooFunction'),
        Value=function_to_export.ref(),
    ))
```
So with these lines of code you export the `FooFunction` with the name `ExportFooFunction`. In this way, can I reference them from outside with this name? With the [boto3's](https://boto3.readthedocs.io/en/latest/reference/services/lambda.html#Lambda.Client.invoke) `invoke()` function as well?

> I like using troposphere have a little post-processing script that takes the template and injects extra resources into them (or updates them).

Uooo great tool, I have been using a bash script for template post-processing so far, thanks for the suggestion.

**Comment by stealthycoin** [comment]
Yea so in another cloudformation template you can use `{"Fn::Import": "ChaliceFooFunction"}` to get a reference to the function. Docs are [here](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-importvalue.html) for it.

Now if you want to just reference it in arbitrary code its a little more difficult. You need to know the name of the output and the name of the stack that it was deployed in, then you can use boto3 or the CLI or whatever SDK to get its real name you would do something like this:

```python
import boto3


def _get_function_from_stack(function_output_key, stack_name):
    cfn = boto3.client('cloudformation')
    result = cfn.describe_stacks(StackName=stack_name)
    outputs = result['Stacks'][0]['Outputs']
    for output in outputs:
        if output['OutputKey'] == function_output_key:
            return output['OutputValue']
    return None


if __name__ == '__main__':
    print(_get_function_from_stack('ExportFooFunction', 'ExampleStack'))
```
Note I didn't test this at all its just for illustration.
